src/Polysemy/Internal.hs

```diff
@@ -534,6 +540,143 @@ subsumeUsing pr =
 {-# INLINE subsumeUsing #-}
 
 
+------------------------------------------------------------------------------
+-- | Finds the effect @e@ in the stack and inserts the effects @es@ after it.
+--
+-- @since TODO
+type family InsertedAfter (e :: Effect) (es :: EffectRow) (r :: EffectRow) :: EffectRow where
+  InsertedAfter e es (e : r) = e : Append es r
+  InsertedAfter e es (e0 : r) = e0 : InsertedAfter e es r
+
+------------------------------------------------------------------------------
+-- | Introduce a set of effects into 'Sem' after the position of the provided
+-- effect @e@. This is intended to be used with a type application:
+--
+-- @
+-- let
+--   sem1 :: Sem [e1, e2, e3, e4, e5] a
+--   sem1 = insertAfter @e2 @[e3, e4] (sem0 :: Sem [e1, e2, e5] a)
+-- @
+--
+-- @since TODO
+insertAfter ::
+  ∀ e es r r' a .
+  r' ~ InsertedAfter e es r =>
+  Subsume r r' =>
+  Sem r a ->
+  Sem r' a
+insertAfter = subsume_
+{-# inline insertAfter #-}
+
+------------------------------------------------------------------------------
+-- | Finds the effect @e@ in the stack and inserts the effects @es@ before it.
+--
+-- @since TODO
+type family InsertedBefore (e :: Effect) (es :: EffectRow) (r :: EffectRow) :: EffectRow where
+  InsertedBefore e es (e : r) = Append es (e : r)
+  InsertedBefore e es (e0 : r) = e0 : InsertedBefore e es r
+
+------------------------------------------------------------------------------
+-- | Introduce a set of effects into 'Sem' before the position of the provided
+-- effect @e@. This is intended to be used with a type application:
+--
+-- @
+-- let
+--   sem1 :: Sem [e1, e2, e3, e4, e5] a
+--   sem1 = insertBefore @e5 @[e3, e4] (sem0 :: Sem [e1, e2, e5] a)
+-- @
+--
+-- @since TODO
+insertBefore ::
+  ∀ e es r r' a .
+  r' ~ InsertedBefore e es r =>
+  Subsume r r' =>
+  Sem r a ->
+  Sem r' a
+insertBefore = subsume_
+{-# inline insertBefore #-}
```

Comment:
Given the problems I pointed out with `insertBefore` and `insertAfter`, my opinion remains that we shouldn't include them.

Reply:
That is, `insertBefore` and `insertAfter` plays hell with even partially polymorphic effects. This doesn't work:
```
insertAfterTest :: forall i r a. Sem (Reader Int ': Reader i ': r) a -> Sem (Reader Int ': Reader i ': State Int ':  r) a
insertAfterTest = insertAfter @(Reader i) @'[State Int]
```
This is unavoidable as long as we try to use effect equality in order to determine where the new effects should be inserted, making `insertAfter` and `insertBefore` fundamentally flawed.